### ISSUE TITLE
MPS] torch.nanmedian SIGABRT on ndim > 4 with dim=0 (MPSNDArraySort axis limit)

### ISSUE THREAD

**Issue by christianahmed905** [pytorch/pytorch#187017]
### 🐛 Describe the bug

"""
torch.nanmedian — SIGABRT (process crash) on MPS for high-rank inputs

torch.nanmedian crashes the Python process with SIGABRT (exit code -6) on MPS
when the input tensor has ndim > 4 and the reduction axis maps to position
>= 4 from the right (e.g. dim=0 on a 5D tensor).  No Python exception is
raised; the process is killed and cannot be recovered.

Crash origin:
  MPSNDArraySort.mm:252: failed assertion `(null)' Axis = 4.
    This class only supports axis = 0, 1, 2, 3

Root cause: torch.nanmedian on MPS uses the same MPSNDArraySort kernel as
torch.sort.  That kernel right-aligns dimensions and only handles the
rightmost 4 axes.  PyTorch does not validate this constraint before
dispatching to MPS, so high-rank tensors with dim in the "left" portion
trigger a fatal assertion inside the Metal runtime.

General crash rule:  crash when  dim < ndim - 4  (i.e. axis from right >= 4)

Failing example (fuzz crash):
  shape=(2, 1, 3, 1, 2)  rank=5  dim=0  keepdim=True  dtype=int64

Crash statistics (Specialized generator, mps):
  torch.nanmedian — 1 crash (hash [PROC], signal=6)
                    example: shape=(2, 1, 3, 1, 2) dim=0 keepdim=True
                             dtype=int64 has_nan=False

Expected: (values, indices) tensors, OR a catchable RuntimeError (e.g.
          "MPS supports at most 4 dimensions for this operation"), or CPU
          fallback.
Actual:   SIGABRT kills the Python process — not catchable with try/except.

Related:  same MPSNDArraySort limit affects torch.sort / torch.argsort
          (see mps_high_rank_sigabrt_20260515.py).  torch.median may also
          be affected.
Affected op:     torch.nanmedian (dim-form reduction)
Affected device: mps (Apple Silicon)
Not affected:    ndim <= 4 on MPS; same inputs on CPU
Versions:        PyTorch 2.12.0, macOS 26.4.1 arm64 (Apple Silicon MPS)
"""

import platform
import subprocess
import sys

import torch

SEP = "-" * 72
CRASH_SHAPE = (2, 1, 3, 1, 2)
CRASH_DIM = 0
CRASH_KEEPDIM = True
CRASH_DTYPE = torch.int64

def _header(title: str) -> None:
    print(f"\n{SEP}\n  {title}\n{SEP}")

def _run_sigabrt_repro() -> None:
    """Run the failing call in a child process (SIGABRT kills the interpreter)."""
    code = f"""
import torch
x = torch.randint(0, 10, {CRASH_SHAPE}, dtype=torch.int64, device="mps")
torch.nanmedian(x, dim={CRASH_DIM}, keepdim={CRASH_KEEPDIM})
"""
    proc = subprocess.run([sys.executable, "-c", code], capture_output=True, text=True)
    if proc.returncode in (-6, 134):
        err = (proc.stderr or proc.stdout or "").strip()
        line = next(
            (ln for ln in err.splitlines()
             if "failed assertion" in ln or "MPSNDArray" in ln),
            err.splitlines()[-1] if err else "(no stderr)",
        )
        print("  result : ProcessCrash (SIGABRT, signal=6)")
        print(f"  origin : {line}")
        return

    if proc.returncode == 0:
        print("  result : UNEXPECTED — call succeeded (no crash on this build)")
        return

    print(f"  result : exit code {proc.returncode}")
    if proc.stderr:
        print(f"  stderr : {proc.stderr.strip()}")

def main() -> None:
    print(SEP)
    print("  torch.nanmedian — MPS high-rank SIGABRT repro")
    print(SEP)
    print(f"  PyTorch  : {torch.__version__}")
    print(f"  Platform : {platform.system()} {platform.machine()}")

    if not torch.backends.mps.is_available():
        print("\n  MPS not available — this repro requires Apple Silicon.")
        sys.exit(1)

    device = torch.device("mps")
    print(f"  Device   : {device}")

    # ── 1. Safe cases on MPS ──────────────────────────────────────────────────
    _header("1. Safe cases (MPS — should succeed)")

    x4 = torch.randint(0, 10, (2, 4, 8, 8), dtype=CRASH_DTYPE, device=device)
    out4 = torch.nanmedian(x4, dim=0, keepdim=True)[0]
    print(f"  ndim=4, dim=0        shape {tuple(x4.shape)}  ->  {tuple(out4.shape)}")

    x5_ok = torch.randint(0, 10, CRASH_SHAPE, dtype=CRASH_DTYPE, device=device)
    out5 = torch.nanmedian(x5_ok, dim=1, keepdim=True)[0]
    print(f"  ndim=5, dim=1 (safe) shape {tuple(x5_ok.shape)}  ->  {tuple(out5.shape)}")

    # ── 2. Crash repro on MPS ─────────────────────────────────────────────────
    _header("2. Crash repro (MPS — expect SIGABRT)")

    print(f"  input    : shape={CRASH_SHAPE}  dtype={CRASH_DTYPE}")
    print(f"  call     : torch.nanmedian(x, dim={CRASH_DIM}, keepdim={CRASH_KEEPDIM})")
    print("  note     : dim=0 on 5D maps to axis 4 from the right (over MPS limit)")
    _run_sigabrt_repro()

    # ── 3. CPU reference ──────────────────────────────────────────────────────
    _header("3. CPU reference (same inputs — should succeed)")

    x_cpu = torch.randint(0, 10, CRASH_SHAPE, dtype=CRASH_DTYPE)
    vals, idx = torch.nanmedian(x_cpu, dim=CRASH_DIM, keepdim=CRASH_KEEPDIM)
    print(f"  values   : shape {tuple(vals.shape)}")
    print(f"  indices  : shape {tuple(idx.shape)}")

    print(f"\n{SEP}\n  done\n{SEP}\n")

if __name__ == "__main__":
    main()

### Versions

PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.5 (main, May 10 2026, 19:20:57) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[conda] Could not collect

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by Isalia20** [comment]
Thanks for the report, reproduced on nightly. Will fix
